### ISSUE TITLE
Navigating from 404 page errors on production deployment

### ISSUE THREAD

**Issue by gtfargo** [gatsbyjs/gatsby#1838]
When a user attempts to navigate _from_ the 404 page on a production build, the console shows this error and the page never redirects:
```
Uncaught TypeError: Cannot read property 'push' of undefined
    at eventHandler (production-app.js:51)
    at mitt.js:1
    at Array.map (<anonymous>)
    at Object.emit (mitt.js:1)
    at done (loader.js:284)
    at loader.js:295
    at loader.js:95
    at loader.js:78
    at index.js?d7c6:11
    at handler (patch.js:18)
```

or, if content for the desired destination is already loaded: 
```
Cannot read property 'push' of undefined
    at navigateTo (production-app.js:66)
    at t.navigateTo (index.js:154)
    at gatsby-browser.js:12
    at catch-links.js:51
```

**404.js**
```
import React from 'react';
import Link from 'gatsby-link';

const NotFoundPage = () =>
  <div>
    <h1>NOT FOUND</h1>
    <Link to={`/`}>Click here to go home</Link>
  </div>;

export default NotFoundPage;
```

**dependency versions:**
```
"gatsby": "^1.8.11",
"gatsby-link": "^1.6.7",
"gatsby-plugin-catch-links": "^1.0.4",
```

**Comment by kbariotis** [comment]
I have the same error using a `gatsby-link` at my [404 component](https://github.com/kbariotis/kostasbariotis.com/blob/master/src/pages/404.js).

**Comment by gtfargo** [comment]
Attempted to reproduce with a the boilerplate example... Seems my personal deployment is causing the issue.  Still digging into this for more info.

**Comment by gtfargo** [comment]
Seems the most recent update has fixed this, @kbariotis try updating.

**Comment by kbariotis** [comment]
@gtfargo hmm not really, still [the same issue](https://kostasbariotis.com/a-non-existing-page).

**Comment by mikhail-softgrad** [comment]
I have same issue

**Comment by KyleAMathews** [comment]
One of you want to take a crack at fixing this?

**Comment by kbariotis** [comment]
@KyleAMathews Sure I will. Could you please guide me to the right direction in the codebase, since I haven't touched it before?

**Comment by gtfargo** [comment]
In the if statement of this eventHandler in `production-app.js` the event value is _only_ the `to` value of the gatsby-link.  In my case, `"/"`

```
function eventHandler(e) {
      if (e.page.path === loader.getPage(pathname).path) {
        emitter.off(`onPostLoadPageResources`, eventHandler)
        clearTimeout(timeoutId)
        window.___history.push(pathname)
      }
    }
```

**Comment by KyleAMathews** [comment]
The error is happening in https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/cache-dir/production-app.js

Follow the [contributing instructions](https://www.gatsbyjs.org/docs/how-to-contribute/#contributing) so you can develop against the checked out version of Gatsby and start adding console.logs to understand why there's the failure.

**Comment by attfarhan** [comment]
@KyleAMathews @kbariotis @gtfargo I also just ran into this issue. Anything we can do to get the PR updated and merged soon?

**Comment by kbariotis** [comment]
Im sorry @attfarhan didn't find the time to work on it :/ I need it too so I will try to check it again soon.

**Comment by revolunet** [comment]
same here; some more info; quite critical because once the user lands on a 404, he can never navigate again

<img width="783" alt="capture d ecran 2017-10-06 a 11 47 09" src="https://user-images.githubusercontent.com/124937/31272740-4c222c4a-aa8c-11e7-8b01-2e87b117066a.png">

<img width="273" alt="capture d ecran 2017-10-06 a 11 47 21" src="https://user-images.githubusercontent.com/124937/31272745-4e8c1fcc-aa8c-11e7-98b7-68506019fec4.png">

**Comment by johnryan** [comment]
fwiw this doesn't just affect links, it seems like none of the JS is being loaded in, (i.e. code called in componentDidMount is executed). @KyleAMathews is this a separate issue or is this resolved in your PR?

**Comment by KyleAMathews** [comment]
This is still an issue.

**Comment by emmafallancy** [comment]
@KyleAMathews I can get this done if you don't mind 😃 This particularly happened when there's no history being pushed to `window.___history`

**Comment by KyleAMathews** [comment]
That'd be great!

**Comment by revolunet** [comment]
in the meantime, any idea how to disable client-side on `pages/404.js` only ?

**Comment by emmafallancy** [comment]
@revolunet do you mean as not using the `pages/404.js` page in production?

**Comment by revolunet** [comment]
no i mean, juste disable the client-side javascript on that page so links are clickable :)

btw, i see PR https://github.com/gatsbyjs/gatsby/pull/2667 is getting some love, so it may be better to focus on that real fix :)

**Comment by emmafallancy** [comment]
ah, right 😅 i see what you mean.

**Comment by mickeyreiss-visor** [comment]
Has anyone made meaningful progress on the fix? This is a user-impactful bug.

@KyleAMathews Do you have a sense of where the issue lies? I'd be willing to take a try at the fix.

**Comment by KyleAMathews** [comment]
@mikhail-softgrad this is fixed actually :-) @emmafallancy did several PRs a month or so ago so closing this now.

**Comment by mickeyreiss-visor** [comment]
Thanks, @KyleAMathews!

For others in a similar situation, it looks like @emmafallancy's fix landed via https://github.com/gatsbyjs/gatsby/pull/2667 and was released in `gatsby@1.9.110`.

**Comment by janosh** [comment]
It appears as though this issue was not yet fixed in v2. Can someone confirm this?
